src/Graphs/DirectedGraphBuilder.cs

```diff
@@ -0,0 +1,210 @@
+            {
+                sources.Remove(to);
+            }
+
+            return this;
+        }
+
+        public DirectedGraphBuilder<T> RemoveElement(T element)
+        {
+            if (element == null) throw new ArgumentNullException(nameof(element));
+
+            if (!elements.Remove(element))
+            {
+                return this;
+            }
+            sources.Remove(element);
+
+            foreach (var to in directSuccessors[element])
+            {
+                directPredecessors[to].Remove(element);
+                if (directPredecessors[to].Count == 0)
+                {
+                    sources.Add(to);
+                }
+            }
+            foreach (var from in directPredecessors[element])
+            {
+                directSuccessors[from].Remove(element);
+            }
+
+            return this;
+        }
+
+        public DirectedGraphBuilder<T> RemoveArrow(T from, T to)
+        {
+            if (from == null) throw new ArgumentNullException(nameof(from));
+            if (to == null) throw new ArgumentNullException(nameof(to));
+            if (!elements.Contains(from)) throw new ArgumentException("Element not in graph.", nameof(from));
+            if (!elements.Contains(to)) throw new ArgumentException("Element not in graph.", nameof(to));
+
+            if (!directSuccessors[from].Remove(to))
+            {
+                return this;
+            }
+
+            directPredecessors[to].Remove(from);
+            if (directPredecessors[to].Count == 0)
+            {
+                sources.Add(to);
+            }
+
+            return this;
+        }
+
+        /// <summary>
+        /// Creates an immutable directed graph instance.
+        /// </summary>
+        public IDirectedGraph<T> CreateGraph()
+        {
+            return new AdjacencyListDiGraph<T>(
+                ImmutableHashSet.CreateRange(elements),
+                directSuccessors.ToImmutableDictionary(
+                    pair => pair.Key,
+                    pair => ImmutableHashSet.CreateRange(pair.Value)),
+                directPredecessors.ToImmutableDictionary(
+                    pair => pair.Key,
+                    pair => ImmutableHashSet.CreateRange(pair.Value)));
+        }
+
+        /// <summary>
+        /// Creates an immutable directed acyclic graph instance.
+        ///
+        /// <para>It will verify that no cycles exist.</para>
+        /// </summary>
+        public IDirectedAcyclicGraph<T> CreateAcyclicGraph()
+        {
+            if (isGraphCyclic())
+            {
+                throw new InvalidOperationException("Cannot create an directed acyclic graph with cycles.");
+            }
+
+            return CreateAcyclicGraphUnsafe();
+        }
+
+        private bool isGraphCyclic()
+        {
+            var visited = new HashSet<T>();
+            var currentStack = new HashSet<T>();
+
+            return elements.Where(checkGraphCycleUsingDepthFirst).Any();
```

Comment:
Isn't this more easy to understand as a regular loop(, or maybe recursive call)?

Reply:
Definitely not a regular loop. Implementing depth-first using a regular loop will only become messy. The recursive call is actually what happens in the method below. I could pull it out, but I don't think the logic would change whatsoever. We would just pass in `visited` and `currentStack` as parameters instead. Feel free to suggest an alternative solution though.

Reply:
After stepping through it, I see why the complexity of the algorithm is necessary. I tried writing a version without `currentStack`, but it is needed to not interpret converging paths as cycles while also being optimal (O(n+m)).

An O(n log n + m) algorithm can be a lot prettier (sort elements by number of predecessors in a priority queue, repeatedly dequeue the top if it has no predecessors, until the priority queue is empty, or the top element has a predecessor - i.e. topologically sorting the graph), but it _is_ less efficient. :(

Reply:
I still think this can be made more readable. I'll write several proposals.

Reply:
Took part of the changes